Validate Sparse Compressed tensor inputs

pearu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79385

The validation includes regular tensor inputs, batched tensor inputs, as well as hybrid tensor inputs.

pearu committed:
Validate Sparse Compressed tensor inputs

The validation includes regular tensor inputs, batched tensor inputs, as well as hybrid tensor inputs.

[ghstack-poisoned]

pearu committed:
Update on "Validate Sparse Compressed tensor inputs"


The validation includes regular tensor inputs, batched tensor inputs, as well as hybrid tensor inputs.

[ghstack-poisoned]

pearu committed:
Update on "Validate Sparse Compressed tensor inputs"


The validation includes regular tensor inputs, batched tensor inputs, as well as hybrid tensor inputs.

[ghstack-poisoned]

pearu committed:
Update on "Validate Sparse Compressed tensor inputs"


The validation includes regular tensor inputs, batched tensor inputs, as well as hybrid tensor inputs.

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/SparseCsrTensorUtils.h:
```diff
@@ -181,6 +181,38 @@ inline std::string plainIndicesName(Layout layout) {
+  }
+}
+
+inline std::string plainDimName(Layout layout) {
+  switch (layout) {
+    case kSparseCsr:
+      return "column";
+    case kSparseCsc:
```
How is this function different from `compressedDimName`?

nikitaved commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -58,199 +59,206 @@ namespace {
+  tensor factory functions.
+
+  The CSR and BSR invariants for PyTorch are outlined in
+
+    https://pearu.github.io/csr_tensor_invariants.html
```
It mentions that `int32` is default for CSR, which is not the case for COO. I have also not seen any indexing tensors which are not `int64` in the dense case.

pearu replied:
https://pearu.github.io/csr_tensor_invariants.html is a little out-of-date: `int64` is the current default for CSR tensors.

The `int32` index support originates from `MKL_INT` being 32-bit integer, see [comment](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/SparseCsrTensorImpl.h#L21-L23).

nikitaved commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -58,199 +59,206 @@ namespace {
-        return 2;
-      });
-  int dense_ndim = values.dim() - compressed_indices.dim() - block_ndim;
-  TORCH_CHECK(dense_ndim == 0, "non-zero dense dimensions (=", dense_ndim, ") is not supported for ", layout, " layout");
+              batch_ndim >= 0,
+              compressed_indices_name, " must have dimensionality >= 1 but got ", compressed_indices.dim());
```
Shouldn't the check be `compressed_indices.dim() == batch_ndim + 1`?

nikitaved commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -58,199 +59,206 @@ namespace {
   AT_DISPATCH_INDEX_TYPES(compressed_indices_type, "validate_sparse_compressed_tensor_args",
-      [&] {
+    [&] {
+      if (plain_indices.numel() > 0) {
         Tensor compressed_indices_cpu = compressed_indices.to(kCPU);
+        Tensor plain_indices_cpu = plain_indices.to(kCPU);
+        int64_t batch_compressed_stride = compressed_indices_cpu.dim() >= 2 ? compressed_indices_cpu.stride(-2) : 0;
+        int64_t batch_plain_stride = plain_indices_cpu.dim() >= 2 ? plain_indices_cpu.stride(-2) : 0;
         auto compressed_indices_data_ptr = compressed_indices_cpu.data_ptr<index_t>();
-        auto batch_stride = compressed_indices_cpu.dim() >= 2 ? compressed_indices_cpu.stride(-2) : 0;
-        auto compressed_dims = (block_ndim == 0 ? size[compressed_dim] : size[compressed_dim] / blocksize[compressed_dim - batch_ndim]);
+        auto plain_indices_data_ptr = plain_indices_cpu.data_ptr<index_t>();
```
From here and below: it is much more efficient to check the invariance with the TensorIterator, i.e. using `cpu_kernel` and `gpu_kernel`, without memory copies and syncs.

pearu replied:
Sure. We should implement this optimization as a follow-up.

In general, validating sparse compressed tensor inputs is (and remains) an expensive operation because one needs to access indices data. Internal tools should use the unsafe constructors for creating such tensors for maximum performance while the correctness of inputs should be checked at the unit tests level.

nikitaved commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -58,199 +59,206 @@ namespace {
       ") must be CPU or CUDA");
-
+  // 4.2, 4.3, 4.4
+  TORCH_CHECK(
+      compressed_indices.get_device() == values.get_device(),
+      "device of ", compressed_indices_name, " (=",
+      compressed_indices.device(),
+      ") must match device of values (=",
+      values.device(),
+      ")");
+  TORCH_CHECK(
+      compressed_indices.get_device() == plain_indices.get_device(),
+      "device of ", compressed_indices_name, " (=",
+      compressed_indices.device(),
+      ") must match device of ", plain_indices_name," (=",
+      plain_indices.device(),
+      ")");
 }
```
A helper function could be used here to eliminate duplicates.

pearu replied:
I am not sure if a helper function is worth implementing here considering that it would be the only use case of such a function and the outputs of these checks are fully tested in
https://github.com/pytorch/pytorch/blob/384ae783c478040bd9ec5142e648786b10598a52/test/test_sparse_csr.py#L741-L759
What do you think, @nikitaved?

nikitaved replied:
Makes sense. Fine by me :)

nikitaved commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -360,34 +368,54 @@ DimVector _estimate_sparse_compressed_tensor_size(
+  const int batch_ndim = compressed_indices.dim() - 1;
+  const std::string compressed_indices_name = compressedIndicesName(layout);
+  const std::string plain_indices_name = plainIndicesName(layout);
+  TORCH_CHECK(
+              batch_ndim >= 0,
+              compressed_indices_name, " must have dimensionality >= 1 but got ", compressed_indices.dim());
+  TORCH_CHECK(
+              compressed_indices.dim() == plain_indices.dim(),
+              compressed_indices_name, " and ", plain_indices_name, " dimensionalities must be equal but got ",
```
From here and below: don't we have all these checked in `_validate_sparse_compressed_tensor_args_worker`? Could this method be called before `_validate_sparse_compressed_tensor_args_worker`?

pearu replied:
Yes, these checks are run twice when a user does not specify the tensor size: in `_estimate_sparse_compressed_tensor_size` followed by `_validate_sparse_compressed_tensor_args_worker` call. However, we'll need these checks here to make sure that the inputs to `_estimate_sparse_compressed_tensor_size` are valid similar to `_validate_sparse_compressed_tensor_args_worker` that assumes a specified size. Notice that the overhead of these checks is minimal (it is about comparing two integers), so, performance-wise, it should be ok.

No, `_estimate_sparse_compressed_tensor_size` cannot be called before `_validate_sparse_compressed_tensor_args_worker` in general because (i) when an user specifies the `size`, there is no need to estimate the size from inputs, and (ii) the sizes returned by `_estimate_sparse_compressed_tensor_size` and specified by the user may be different while both being valid sizes (`_estimate_sparse_compressed_tensor_size` returns a *minimal* valid size for the given input).

nikitaved commented on test/test_testing.py:
```diff
@@ -1308,7 +1308,7 @@ def test_matching(self):
 
     def test_mismatching_ccol_indices_msg(self):
         actual_ccol_indices = (0, 1, 2)
-        actual_row_indices = (1, 0)
```
Otherwise it fails with unsorted indices, right?

pearu replied:
Yes, correct. This changeset fixes incorrect test data.

nikitaved commented on test/test_sparse_csr.py:
```diff
@@ -171,46 +171,131 @@ def genTensor(self, size, nnz, *, layout, device=None, dtype=torch.float, index_
+                if not lst:
+                    return torch.tensor(lst, device=device, dtype=dtype).reshape(0, *blocksize)
+
+                def list_add(lst, value):
+                    if isinstance(lst, list):
+                        return [list_add(item, value) for item in lst]
+                    return lst + value
```
Can you have list of lists?

pearu replied:
yes, this is exercised in BSR/BSC-related tests.

nikitaved commented on test/test_sparse_csr.py:
```diff
@@ -502,6 +574,274 @@ def test_consistency(self, layout, device, dtype, op):
+
+    def _generate_invalid_input(self, layout):
+
+        def shape(shape, basedim=0, blocksize=(1, 1)):
+            # Below, we define compressed and plain indices that
```
This function is already defined above? Or are they not identical?

pearu replied:
Fixed via https://github.com/pytorch/pytorch/pull/79385#discussion_r907200124

nikitaved commented:
@cpuhrsch , it seems like after this PR is merged, we will be able to construct hybrid compressed tensors. What would be the minimal functionality, when implemented, to be sufficient to say that dense dimensions are supported? `resize` and `to_dense`?

cpuhrsch commented:
@nikitaved - An extension of `to_sparse_csr` akin to the native function [`to_sparse.sparse_dim`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L5879-L5883) so that users can convert from, at least, the strided layout to this layout and a conversion back to strided via `to_dense`. I don't think `resize` is necessary right away, but what makes you suggest it?

cpuhrsch reviewed (commented):
I think generally this looks good (the large number of changes here are due to the expect files), but @nikitaved I raising good points about various details. Let's wait until Pearu is back from his PTO next week to further process this.

pearu commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -58,199 +59,206 @@ namespace {
-                        "(Batch element ", batch_id, ") ",
-                        "at position i = ", i, ", the condition ", compressed_indices_name, "[i - 1] <= ", compressed_indices_name, "[i] fails, got ",
-                        compressed_indices_data_ptr[batch_id*batch_stride + i - 1], " <= ", compressed_indices_data_ptr[batch_id*batch_stride + i]);
+          int64_t start = compressed_indices_data_ptr[batch_id*batch_compressed_stride];
+          const std::string at_batch_id = (batch_ndim > 0 ? " at batch id " + std::to_string(batch_id) : "");
+          const std::string batch_indices = (batch_ndim > 0 ? "..., " : "");
+          // 5.1
```
I suggest doing this optimization in a follow-up because this PR targets correctness rather than efficiency, see the general note in https://github.com/pytorch/pytorch/pull/79385#discussion_r907159974

nikitaved replied:
I agree, and I am currently working on it.

pearu commented on test/test_sparse_csr.py:
```diff
@@ -502,6 +574,274 @@ def test_consistency(self, layout, device, dtype, op):
+                if not lst:
+                    return torch.tensor(lst, device=device, dtype=dtype).reshape(0, *blocksize)
+
+                def list_add(lst, value):
+                    if isinstance(lst, list):
+                        return [list_add(item, value) for item in lst]
+                    return lst + value
```
Fixed: introduced `_generate_small_inputs_utils` method that returns `shape` and `values` utility functions.

pearu committed:
Update on "Validate Sparse Compressed tensor inputs"


The validation includes regular tensor inputs, batched tensor inputs, as well as hybrid tensor inputs.

[ghstack-poisoned]

pearu committed:
Update on "Validate Sparse Compressed tensor inputs"


The validation includes regular tensor inputs, batched tensor inputs, as well as hybrid tensor inputs.

[ghstack-poisoned]

nikitaved reviewed (approved):
Thank you, @pearu ! LGTM.

pearu committed:
Update on "Validate Sparse Compressed tensor inputs"


The validation includes regular tensor inputs, batched tensor inputs, as well as hybrid tensor inputs.

[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot merge
